### ISSUE TITLE
DataTable: Provide screen reader compliant keyboard navigation

### ISSUE THREAD

**Issue by Rene-Stoeckel** [primefaces/primefaces#11783]
### Description

One can use the keyboard to navigate between the rows by using the up and down 
arrow keys once the focus is moved from the table headers to the table body,
as demonstrated in the showcase here: 

http://www.primefaces.org:8080/showcase/ui/data/datatable/lazy.xhtml

But this approach is insufficient in combination with screan readers (e.g. JAWS) 
and thus is a serious accessibility limitation.

The problem is that the data tables are given the aria role "grid" so that 
screen readers expect the "W3C Pattern Grid Interactive Tabular Data" (see below)
and expect the table to be navigated via the cursor keys from row to row and from cell to cell 
similiar to excel, which is not implemented.
The focus currently stays on top on tbody instead of the "cells" so that the screen 
reader does not read anything from the table content during navigation.

### Describe the solution you would like

1) It would be nice to have an implementation of table navigation from row to row and cell to cell 
via cursor keys/ home / end / page up / page down according to the example in the WAI Pattern:
https://www.w3.org/WAI/ARIA/apg/patterns/grid/examples/data-grids/

2) Instead of staying on tbody the keyboard focus would need to follow the cursor 
so that the screen readers will read the cells content.

### Additional context

This is the heaviest finding in the acessibility assessment of our primefaces application.

Thank you for your great work on primefaces!

**Comment by Rene-Stoeckel** [comment]
We are currently using Primefaces 13.0.1 / 13.0.7 with browser edge as primary and chrome/firefox as secondary.

**Comment by melloware** [comment]
PR is welcome.  Datatable is quite a complex beast.

**Comment by Rene-Stoeckel** [comment]
PR is currently no option, we don't have the capacity at the monent. Is there a way to solve this issue via sponsoring or commercial license?

**Comment by melloware** [comment]
Yes see PRO support and click Get A Quote http://www.primefaces.org/showcase/support.xhtml

**Comment by melloware** [comment]
@Rene-Stoeckel i moved this ticket under sponsoring required: https://github.com/primefaces/primefaces/issues/11251

**Comment by melloware** [comment]
@Rene-Stoeckel i sent you an email.

**Comment by melloware** [comment]
@Rene-Stoeckel another email sent do you not check your email regularly?

**Comment by melloware** [comment]
@Rene-Stoeckel please have your team test out this reproducer:
[pf-11918.zip](https://github.com/primefaces/primefaces/files/15336327/pf-11918.zip)

Unzip and run `mvn clean jetty:run` and navigate to: http://localhost:8080/

Now cells are focuses and you can use Arrow keys and all of the keystrokes mentioned in the doc to navigate the cells and screen reader should read them.

![image](https://github.com/primefaces/primefaces/assets/4399574/22af45e7-c641-4622-ac73-34de1a38a6b1)



![datatable-navigation](https://github.com/primefaces/primefaces/assets/4399574/50fb2ce4-c065-4f97-bbfc-5d090fcf4eaa)

**Comment by Rene-Stoeckel** [comment]
We just tested the reproducer (from 11918) with the screenreader and we can confirm that

1) the content of the table is correctly presented with the screenreader, just as expected.
2) cursor navigation is working as expected,

But there are things we noticed that still must be fixed here:

When a cell other than (1,1) is clicked with the mouse the cursor navigation is broken. The fokus travels to a different cell than expected and some cells can no longer be reached. E.g. select cell New York, navigate all way down and then all way up (cursor or page down/up) and the fokus will end up in cell 123.

**Comment by melloware** [comment]
Try this one:
[pf-dt-accessibility.zip](https://github.com/primefaces/primefaces/files/15349155/pf-dt-accessibility.zip)

**Comment by Rene-Stoeckel** [comment]
Thank you very much for the quick response. We tried this tester and the two issues are fixed. Screenreade and cursor navigation are as expeced.

**Comment by Rene-Stoeckel** [comment]
Is there a way to test the screen reader compliant keyboard navigation together with the data tables row expansion feature, to ensure fitness for overall integration?

**Comment by melloware** [comment]
You should be able to test that here: http://20.161.11.140/showcase/ui/data/datatable/expansion.xhtml

Be warned though the Datatable is a COMPLEX beast so supporting every scenario will be difficult.  I disable the cell navigation on RowSelection examples only because RowSelection hsa its own keyboard handling and refactoring it all would be a HUGE change.

**Comment by melloware** [comment]
Looks like I have a fix to make so SPACE/ENTER expands and collapses the row.  Let me know if you see anything else

**Comment by Rene-Stoeckel** [comment]
Yes, I see, it's not quite trivial. The expand via space/enter would need a fix. Also expanding a row inserts a new row <tr class="ui-expanded-row-content ui-widget-content"> which the keyboard navigation can not yet handle or to get over it to the rows below. When the row is unexpanded, it also leaves the tr there, as invisible, with  style="display: none;"> which still irritates the row-up, row-down navigation logic.

**Comment by melloware** [comment]
let me see what I can do!

**Comment by melloware** [comment]
@Rene-Stoeckel feel free to re-test i have fixed the SPACE/ENTER triggering the row toggler and the ARROW keys work to get around the new expanded row.

http://20.161.11.140/showcase/ui/data/datatable/expansion.xhtml

**Comment by Rene-Stoeckel** [comment]
Our team was able to take a closer look onto the demo again and here is the feedback. 
Although we face a complex topic, the feedback here is not that bad though:

The datatable is attributed with role=grid. As a consequence the users of the screenreader expects to be able to reach all cells (and assiciated data) with the cursor keys and that the tab key is used to leave the grid (according to [Grid (Interactive Tabular Data and Layout Containers) Pattern | APG | WAI | W3C](https://deu01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fwww.w3.org%2FWAI%2FARIA%2Fapg%2Fpatterns%2Fgrid%2F&data=05%7C02%7CRene.Stoeckel%40telekom.de%7C2ebe59e1cc014250eba608dc83b5670e%7Cbde4dffc4b604cf68b04a5eeb25f5c4f%7C0%7C0%7C638530062348354389%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C0%7C%7C%7C&sdata=h6LW9QqxVb7P%2FfjcgnpBGL0REUG%2Fis9tXI%2FPj8SBLc8%3D&reserved=0)).


Currently it is not fully implemented that way:
•	The expanded rows cannot yet be reached with the cursor keys. It is not possible to use the screen reader in linear reading mode here either because the screenreader (as intended) switches to the form mode.
•	The tab key cannot be used to leave the table, as it navigates to the next (row expansion) buttons within the grid.

Also, it is unclear, why the button in the first row are labeled with numbers beginning from 1000, which takes the screen reader long to articulate.

When the two issues mentioned above (usability with cursor keys and TAB) would be solved then the remaining issues would just be minor priority (and it would pass the accessability acceptance in priciple).

**Comment by melloware** [comment]
@Rene-Stoeckel good questions again.

1. for the Collapse/Expand ARIA label its because the toggler does this.

```js
var ariaLabel = expanded ? PrimeFaces.getAriaLabel('collapseLabel') : PrimeFaces.getAriaLabel('expandLabel');
            if (rowMeta && rowMeta.key) {
                ariaLabel += " " + rowMeta.key;
            }
```

basically it uses whatever your unique "rowKey" is to identify the expand/collapse button.  So it will read "Expand Row 1000" or whatever the rowkey of the row is because if not it would just read "Expand Row" but what row???

2. As for the TAB key obviously its because row togglers have `tabIndex="0"` so they can be tabbed to.  I don't think ARIA guidelines have any examples for expandable tables so there is really no guidelines to go against.  My thought is the need to remain tabbable as they are buttons in the table.

Thoughts?

**Comment by Rene-Stoeckel** [comment]
We are still in discussion how to handle this case (accessibility for datatables with row expansion). On the one hand we have a clear recommendation to stick with the standard accessibility patterns (in this case data grid), on the other hand we are at the moment to some degree dependent on the row expansion feature.
